Introduce the bundle storage fund

NingLin-P opened the pull request:
This PR introduces the bundle storage fund, which is used to pay the bundle storage fee when a bundle is submitted for the operator:
- The bundle storage fund is an account that derives from the operator id (unique & per operator)
- The bundle storage fee is paid by the fund when the operator submits the bundle and the storage fee collected from the domain user will be refunded back to the operator when the ER is confirmed
  - For a given domain block, the amount refund to a particular operator is determined by the total storage fee collected from the domain user and the percentage of bundle storage that the operator has submitted for the domain block: `total_storage_fee * operator_submitted_bundle_size / total_bundle_size`

The deposit and withdrawal of the bundle storage fund are unimplemented in this PR, they are based on staking thus depending on #2390, so in this PR the charge and refund of the bundle storage fee is disabled. 

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

NingLin-P committed:
Introduce the StorageFeeInterface and the PalletId for pallet-domains

The StorageFeeInterface is used to query the transaction_byte_fee and collect the
charged bundle storage fee to the pallet-transaction-fees, which will reward the
fee to the block author later

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Collect the submitted bundle size

The bundle size info will be used to refund storage fee to the operator according
to the total size of bundle they have submitted.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Introduce bundle storage fund

This fund is derive from the operator and used to pay the bundle storage fee
when the operator submit the bundle and refund the storage fee that collected
from the domain user back to the operator when the ER is confirmed

Signed-off-by: linning <linningde25@gmail.com>

dariolina commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -1425,6 +1463,20 @@ mod pallet {
+                    ) {
+                        log::debug!(
+                            target: "runtime::domains",
+                            "Operator {} unable to pay for the bundle storage fee, domain id {:?}, error: {e:?}",
```
It would be nice to have the operator only include as many tx into the bundle as he can pay for. Not a blocker, this current impl is in line with the spec.

NingLin-P replied:
Leave a TODO for now as this requires getting the `transaction_byte_fee` from the consensus chain and is duplicated with #2384

NingLin-P committed:
Minor test fixes and TODO

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Only update the bundle bundle_size and estimated_bundle_weight iff the extrinsic is added to the bundle body

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into operator-storage-fee

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -259,6 +260,8 @@ pub(crate) struct ConfirmedDomainBlockInfo<DomainNumber, Balance> {
     pub rewards: Balance,
     pub invalid_bundle_authors: Vec<OperatorId>,
+    pub total_storage_fee: Balance,
+    pub front_paid_storage: BTreeMap<OperatorId, u32>,
```
Does `front_paid_storage` means what operator paid during the bundle submission? If so, can we just call it `paid_storage_fee` or something more appropriate? I'm not able to relate what `front` here means

NingLin-P replied:
We are getting this when the ER is confirmed so `front` means what the operator has paid previously when submitting the bundle.

vedhavyas replied:
I understand that but naming here is not straight forward unless there is an explanation IMHO `paid_bundle_storage_fees` would be the better representation of the variable.

vedhavyas commented on crates/pallet-domains/src/bundle_storage_fund.rs:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    let storage_fund_acc = storage_fund_account::<T>(operator_id)?;
+    let storage_fee = T::StorageFeeInterface::transaction_byte_fee() * bundle_size.into();
```
```suggestion
    let storage_fee = T::StorageFee::transaction_byte_fee() * bundle_size.into();
```
Can we just call it StorageFee instead ?

vedhavyas commented on crates/pallet-domains/src/bundle_storage_fund.rs:
```diff
@@ -0,0 +1,107 @@
+    let storage_fund_acc = storage_fund_account::<T>(operator_id)?;
+    let storage_fee = T::StorageFeeInterface::transaction_byte_fee() * bundle_size.into();
+
+    T::Currency::burn_from(
```
@dariolina do we need to burn the storage fees here and then mint back again ? Can we instead do a transfer of funds ?
Burning will cause the total issuance to always fluctuate

NingLin-P replied:
This is exactly how every other transaction pays the fee where the tx fee is burnt and noted in `pallet-transaction-fees` then the total tx fee of the current block will be mint into the block author's reward address at block finalization.

vedhavyas replied:
Wouldn't it be better to instead use an protocol controlled account to transfer those fees temporarily and then transfer back to block author instead of burning and there by fluctuating the total supply of SSC in every block. This is out of scope for this PR but I think that would be a better way to handle this on consensus chain. Thoughts @nazar-pc ?

vedhavyas commented on crates/pallet-domains/src/bundle_storage_fund.rs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        let refund_amount = {
+            let share = Perbill::from_rational(paid_storage, total_paid_storage);
```
Maybe we should introduce the same `SharePrice` struct here as well

NingLin-P replied:
We don't have the same issue as the staking share here, we just need to compute a percentage of the total fee for each operator, if `share` is too confusing perhaps rename `share` to `percentage`?

dariolina replied:
FYI @vedhavyas maybe [spec](https://www.notion.so/subspacelabs/Fees-Rewards-Specification-WIP-1b835c7684a940f188920802ca6791f2?pvs=4#6688692ad473489e81c348570300ded6) is a bit more clear. The bundle authors just get refunded proportionally to what they paid in a block. Totally independent of stake shares.

vedhavyas replied:
Yes I understand that but I'm confused with the variable naming here as `share` instead of something like `paid_storage_percentage` or something that not overlapping with staking terminology

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -1555,7 +1607,13 @@ impl<T: Config> Pallet<T> {
+
         ensure!(max_size >= bundle_size, BundleError::BundleTooLarge);
+        ensure!(
+            opaque_bundle.size() == bundle_size,
```
Don't the `opaque_budnlde.size()` includes the ER as well ?

NingLin-P replied:
I don't think so according to the [spec](https://www.notion.so/subspacelabs/Domains-v2-Specification-3fb0ec6e4d204c4881a7df50ef58da8f?pvs=4#37dc09fa7c604b04a0e79b67adf9c405) and the comment fo this field:
https://github.com/subspace/subspace/blob/6b5ef44bcb7ceb9e2076795c1092cdb19d13b972/crates/sp-domains/src/lib.rs#L210-L213

vedhavyas replied:
hmmm. @dariolina dont we want to also include ER size here as well as this is as well stored

dariolina replied:
So far we have been only talking about the body storage cost, of extrinsics inside it. Currently, yeah, the operators put ERs in basically for free. In this way they are similar to block headers: those are also archived and stored but nobody "paid" for them.
If we make operators pay for ERs storage too we will have to charge users more then. 
I'll bring up this question in R&D, but IMO, it's fine as is.

dariolina replied:
We've discussed this in R&D and the consensus was that ERs should pay, but we will open a separate issue for it to define design and spec. This PR can merge with just the bundle body paying.

vedhavyas commented on crates/sp-domains/src/lib.rs:
```diff
@@ -320,6 +320,11 @@ impl<Extrinsic: Encode, Number: Encode, Hash: Encode, DomainHeader: HeaderT, Bal
+
+    /// Return the bundle body size in bytes
+    pub fn size(&self) -> u32 {
+        self.sealed_header.header.bundle_size
```
hmm shouldn't we include the ER as well ?

vedhavyas commented on crates/subspace-runtime/src/lib.rs:
```diff
@@ -639,6 +639,7 @@ parameter_types! {
     pub const MaxPendingStakingOperation: u32 = 100;
     pub const MaxNominators: u32 = 256;
     pub SudoId: AccountId = Sudo::key().expect("Sudo account must exist");
+    pub const DomainsPalletId: PalletId = PalletId(*b"ssdomain");
```
```suggestion
    pub const DomainsPalletId: PalletId = PalletId(*b"domains_");
```
or 

```suggestion
    pub const DomainsPalletId: PalletId = PalletId(*b"_domains");
```

would be ideal

vedhavyas commented on domains/client/domain-operator/src/tests.rs:
```diff
@@ -1619,6 +1624,11 @@ async fn test_true_invalid_bundles_illegal_extrinsic_proof_creation_and_verifica
         bundle_extrinsic_root =
             BlakeTwo256::ordered_trie_root(extrinsics.clone(), StateVersion::V1);
         opaque_bundle.sealed_header.header.bundle_extrinsics_root = bundle_extrinsic_root;
+        opaque_bundle.sealed_header.header.bundle_size = opaque_bundle
```
maybe introducing `size()` on `opaque_bundle` is a good idea

NingLin-P replied:
Seems we already have a `size()` for returning the `bundle_size` field, the field may not be necessary since the consensus node can't trust this field and has to recalculate the size during `validate_bundle`, perhaps we can remove this field and change `size()` to calculate the size instead, it requires update on the spec cc @dariolina 

Any thoughts?

dariolina replied:
iirc `bundle_size` ER field also doesn't have a fraud proof. If you say it's useless I have no objections to removing it from ER

NingLin-P replied:
This field is attached to the bundle header, not the ER. The consensus node can perform the bundle size check (ensure the bundle size does not exceed the limit) itself thus we don't need fraud proof.

dariolina replied:
Even better. I can remove from spec.

NingLin-P committed:
Add deposit function for the bundle storage fund

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Minor refactoring and rename

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Remove the bundle_size field from the bundle header

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into operator-storage-fee

NingLin-P committed:
Support withdraw from the bundle storage fund

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Adjust staking tests and add test for the bundle storage fund

The staking tests were dejusted due to the bundle storage fund introduce more
calculation and rounding dust

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Enable charge and refund of the bundle storage fee

Signed-off-by: linning <linningde25@gmail.com>

dariolina commented on crates/pallet-domains/src/bundle_storage_fund.rs:
```diff
@@ -0,0 +1,219 @@
+/// Return the bundle storage fund redeem price
+pub fn storage_fund_redeem_price<T: Config>(
+    operator_id: OperatorId,
+    operartor_total_deposit: BalanceOf<T>,
```
```suggestion
    operator_total_deposit: BalanceOf<T>,
```

dariolina commented on crates/pallet-domains/src/bundle_storage_fund.rs:
```diff
@@ -0,0 +1,219 @@
+    operartor_total_deposit: BalanceOf<T>,
+) -> StorageFundRedeemPrice<T> {
+    let total_balance = total_balance::<T>(operator_id);
+    StorageFundRedeemPrice::<T>::new(total_balance, operartor_total_deposit)
```
```suggestion
    StorageFundRedeemPrice::<T>::new(total_balance, operator_total_deposit)
```

dariolina commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -83,29 +83,61 @@ impl From<(DomainId, EpochIndex)> for DomainEpoch {
 
+pub struct NewDeposit<Balance> {
+    pub(crate) staking: Balance,
+    pub(crate) storage_fee: Balance,
```
Can this also be `storage_fee_deposit` like in `KnownDeposit` and `PendingDeposit`?

dariolina commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -83,29 +83,61 @@ impl From<(DomainId, EpochIndex)> for DomainEpoch {
+    pub(crate) domain_id: DomainId,
+    pub(crate) unlock_at_confirmed_domain_block_number: DomainBlockNumber,
+    pub(crate) amount_to_unlock: Balance,
+    pub(crate) storage_fee: Balance,
```
Here too. `storage_fee_refund` maybe? That's what I put in spec.

dariolina reviewed (commented):
Looks good to me except few nits

NingLin-P committed:
Fix typo, rename

Signed-off-by: linning <linningde25@gmail.com>

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -314,7 +317,8 @@ pub(crate) fn process_execution_receipt<T: Config>(
                 ));
 
-                // Collect the invalid bundle author
+                // Collect the front paid storage and the invalid bundle author
```
```suggestion
                // Collect the paid bundle storage fees and the invalid bundle author
```

vedhavyas commented on crates/pallet-domains/src/bundle_storage_fund.rs:
```diff
@@ -0,0 +1,219 @@
+    Ok(())
+}
+
+/// Refund the front paid storage fee of a particular domain block back to the operator, the amount to
```
```suggestion
/// Refund the paid bundle storage fee of a particular domain block back to the operator, the amount to
```

vedhavyas commented on crates/pallet-domains/src/bundle_storage_fund.rs:
```diff
@@ -0,0 +1,219 @@
+
+    let storage_fee_reserve = STORAGE_FEE_RESERVE.mul_floor(deposit_amount);
+
+    T::Currency::transfer(
```
We should also emit an event for this so that this amount can be indexed later and also gives some info to the user

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -79,6 +80,7 @@ pub(crate) type NominatorId<T> = <T as frame_system::Config>::AccountId;
 pub trait HoldIdentifier<T: Config> {
     fn staking_staked(operator_id: OperatorId) -> FungibleHoldId<T>;
     fn domain_instantiation_id(domain_id: DomainId) -> FungibleHoldId<T>;
+    fn storage_fund(operator_id: OperatorId) -> FungibleHoldId<T>;
```
IIUC, this is not the correct representation. We use this hold when withdrawal is enabled. So the hold reason should have a hint for the user. Something like `storage_fund_withdrawal` would be ideal

vedhavyas commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -1872,8 +2121,9 @@ pub(crate) mod tests {
             withdraws: vec![(49 * SSC, Ok(()))],
             maybe_deposit: None,
-            expected_withdraw: Some((49 * SSC, false)),
+            expected_withdraw: Some((48999999980000000000, false)),
```
Why the change in value here. If there is no reward, then withdrawals should be exact. There should not be any float point loss

vedhavyas commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -1886,8 +2136,9 @@ pub(crate) mod tests {
             withdraws: vec![(29 * SSC, Ok(())), (20 * SSC, Ok(()))],
             maybe_deposit: None,
-            expected_withdraw: Some((49 * SSC, false)),
+            expected_withdraw: Some((48999999986852892560, false)),
```
Same here as well

vedhavyas commented on crates/pallet-domains/src/staking.rs:
```diff
@@ -1904,8 +2155,9 @@ pub(crate) mod tests {
             ],
             maybe_deposit: None,
-            expected_withdraw: Some((49 * SSC, false)),
+            expected_withdraw: Some((48999999986852892560, false)),
```
Ditto

vedhavyas reviewed (commented):
The direction of these changes make sense.

I have some concerns over why the values in the tests changed specifically those that should not have any float point loss where there is no reward.  Is this because the storage fees are split between operators while returning the funds back? 

While I'm okay with the change in the types, I would prefer if we also include the necessary migrations for those types in here so that we dont need worry about some changes missing in the other PR

NingLin-P committed:
Add StorageFeeDeposited event, rename

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into operator-storage-fee

NingLin-P committed:
Merge branch 'main' into operator-storage-fee

NingLin-P committed:
Update crates/pallet-domains/src/block_tree.rs

Co-authored-by: Vedhavyas Singareddi <vedhavyas.singareddi@gmail.com>

NingLin-P committed:
Update crates/pallet-domains/src/bundle_storage_fund.rs

Co-authored-by: Vedhavyas Singareddi <vedhavyas.singareddi@gmail.com>

vedhavyas reviewed (approved):
> Right, it is because the deposit is split and transferred from and backward with the storage fund, thus there is some rounding dust generated and left in the storage fund (it is sent to treasury upon unlock/slash).I also added some checks in the test to ensure the total amount of funds is unchanged and just distributed in different places.

hmm. Okay no major concerns but would have been nice to have those dust going to treasury though I dont think this would be a problem practically since operators will produce a bundle anyway.
